### ISSUE TITLE
matcher helper for mutiple http route entry matching

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#20609]
*Title*: *matcher helper for mutiple http route entry matching*

*Description*:

If we have mutiple routes with query match, regex path match, prefix path match, etc, we may call `removeQueryAndFragment`, `parseQueryString`, etc repeatedly. When I have a large number of routes, this may introduces some non-negligible overhead.

May be we can create a simple matcher helper which will lazily create query parameter map, path without query, etc, only once for all routes.

**Comment by zhxie** [comment]
Yes, I agree that we need to remove these overheads. Since the issue is marked as beginner, I may leave to someone new to the community to have a try.

**Comment by wbpcode** [comment]
This may need updates all the match method of route entry and related tests. I am not sure it's easy enough for beginner.

**Comment by zhxie** [comment]
Since `removeQueryAndFragment` and `parseQueryString` only happen to `RequestHeaderMap`, maybe we can just add some lazy methods in `RequestHeaderMap` like `justPath` and `queryParams`?

**Comment by wbpcode** [comment]
I prefer a simple helper to do this, which avoids modifying the RequestHeaderMap API. That's a bigger impact. 🤔

**Comment by zhxie** [comment]
I have evaluated the code path. `removeQueryAndFragment` and `parseQueryString` are majorly called by router which takes the `RequestHeaderMap` for routing and headers modification. Since the router can be used by multiple requests, I think it is impossible to store `path` and `queryParams` in the router itself if we do not want to modify the `RequestHeaderMap`.

**Comment by wbpcode** [comment]
For each request, only once path rewrite is triggered. But if we have mutiple routes, the matching will be triggered repeatly. So I am still inclined to solve the problem in the matching process. In this way, our modifications to the core code will be relatively small, and we can obtain good benefits.
